w12h02/test/pgdp/ds/RingbufferMultithreadTest.java

```diff
@@ -0,0 +1,182 @@
+            for (int i = 0; i < N_REPS; i++) {
+                try{ringBuffer.put(1);
+                ringBuffer.get();}catch(InterruptedException e){}
+            }
+        };
+        Future<?>[] futures = new Future<?>[2];
+        futures[0] = pool.submit(routine);
+        futures[1] = pool.submit(routine);
+        TimeoutChecker t = new TimeoutChecker(1000, futures);
+        assertFalse(t.isTimeoutReached(), "Timeout reached: Deadlock while Putting");
+        assertTrue(ringBuffer.isEmpty());
+    }
+
+    @Test
+    void shouldWaitWhenFull() throws InterruptedException{
+        var pool = Executors.newFixedThreadPool(1);
+        RingBuffer ringBuffer = new RingBuffer(1);
+        ringBuffer.put(1);
+        Runnable routine = () -> {
+            try{ringBuffer.put(1);}catch(InterruptedException e){}
+        };
+        Future<?>[] futures = new Future<?>[1];
+        futures[0] = pool.submit(routine);
+        TimeoutChecker t = new TimeoutChecker(500, futures);
+        pool.shutdown();
+        assertTrue(t.isTimeoutReached(), "Does not wait.");
+        assertTrue(ringBuffer.isFull());
+    }
+
+    @Test
+    void shouldWaitWhenEmpty(){
+        var pool = Executors.newFixedThreadPool(1);
+        RingBuffer ringBuffer = new RingBuffer(1);
+        Runnable routine = () -> {
+            try{ringBuffer.get();}catch(InterruptedException e){}
+        };
+        Future<?>[] futures = new Future<?>[1];
+        futures[0] = pool.submit(routine);
+        TimeoutChecker t = new TimeoutChecker(500, futures);
+
+        assertTrue(t.isTimeoutReached(), "Does not wait.");
+        assertTrue(ringBuffer.isEmpty());
+    }
+
+    @Test
+    void shouldAwakeWhenNotEmptyAnymore() throws InterruptedException {
+        var pool = Executors.newFixedThreadPool(1);
+        RingBuffer ringBuffer = new RingBuffer(1);
+        Runnable routine = () -> {
+            try{ringBuffer.get();}catch(InterruptedException e){}
+        };
+        Future<?>[] futures = new Future<?>[1];
+        futures[0] = pool.submit(routine);
+        TimeoutChecker t1 = new TimeoutChecker(100, futures, false);
+
+        try{Thread.sleep(200);}catch(InterruptedException e){}
+        boolean b = t1.isTimeoutReached(false);
+        System.out.println(b);
+        assertTrue(b, "Does not wait");
+
+        TimeoutChecker t2 = new TimeoutChecker(1000, futures);
+        ringBuffer.put(1);
+        assertFalse(t2.isTimeoutReached(), "Does not proceed wait.");
+    }
+
+    @Test
+    void shouldAwakeWhenNotFullAnymore() throws InterruptedException {
+        var pool = Executors.newFixedThreadPool(1);
+        RingBuffer ringBuffer = new RingBuffer(1);
+        ringBuffer.put(1);
+        Runnable routine = () -> {
+            try{ringBuffer.put(2);}catch(InterruptedException e){}
+        };
+        Future<?>[] futures = new Future<?>[1];
+        futures[0] = pool.submit(routine);
+        TimeoutChecker t1 = new TimeoutChecker(100, futures, false);
```

Comment:
```suggestion
        TimeoutChecker t1 = new TimeoutChecker(100, futures);
```

Reply:
cancelFuturesOnTimeout=false is necessary, because the TimeoutChecker would otherwise cancel it's Futures, so it wouldn't be tested, if they awake after having waited when the Ringbuffer was full/empty.